vmu.py

```diff
@@ -115,3 +90,86 @@ def canonical_src_string(sources):
     return result.strip()
+
+class ParamError(Exception):
+    """Exception for errors in parameter files"""
+    pass
+
+class PackageSet(object):
+    """A class for comparing package sets as specified in VMU test parameters
+
+    Instance parameters:
+
+    label: string describing the package set, used for reporting
+    packages: list of packages as strings
+    selinux: boolean to turn on/off SELinux enforcing mode (default: False)
+    java: boolean to pre-install OSG java packages (default: True)
+
+    PackageSets are equal if their 'packages' lists are the same. If 'packages'
+    lists are equal but 'label' is not, ParamError is raised.
+
+    PackageSets are sorted by labels according to LABEL_ORDER. Unknown labels
+    are sorted alphanumerically after the known labels.
+    """
+
+    OSG_JAVA_DEFAULT = True
+    SELINUX_DEFAULT = False
+    LABEL_ORDER = ['All', 'All + GRAM', 'All + GRAM (3.2)', 'HTCondor', 'GridFTP', 'BeStMan', 'VOMS', 'GUMS']
+
+    def __init__(self, label, packages, selinux=SELINUX_DEFAULT, java=OSG_JAVA_DEFAULT):
+        if not label:
+            raise ParamError("PackageSet 'label' field cannot be blank")
+        if not packages:
+            raise ParamError("PackageSet 'packages' field must be non-empty list")
+        self.label = label
+        self.selinux = selinux
+        self.java = java
+
+        if self.java:
+            self.packages = ['java-1.7.0-openjdk-devel', 'osg-java7-compat', 'osg-java7-devel-compat'] + packages
+        else:
+            self.packages = packages
+
+    def __eq__(self, other):
+        if isinstance(other, PackageSet):
+            if self.packages == other.packages:
+                if self.label != other.label:
+                    raise ParamError("Different package set tags ('%s', '%s') refer to the same set of packages. " %
+                                     (self.label, other.label) + "Check if java attributes are different.")
+                return True
+            elif self.label == other.label:
+                raise ParamError("Package set label '%s' refers to different sets of packages" % self.label)
+            return False
+        return NotImplemented
+
+    def __ne__(self, other):
+        result = self.__eq__(other)
+        if result is NotImplemented:
+            return result
+        return not result
+
+    def __lt__(self, other):
+        def get_sort_val(pkg_set):
+            try:
+                return pkg_set.LABEL_ORDER.index(pkg_set.label)
+            except ValueError:
+                return pkg_set.label
+        return get_sort_val(self) < get_sort_val(other)
+
+    def __hash__(self):
+        # Treat similarly labeled sets of packages as equal for hashing;
+        # SELinux status can differ on a per-run basis
+        return hash(repr(self).replace('%s, ' % self.selinux, ''))
+
+    def __repr__(self):
+        return str([self.label, self.selinux, self.packages])
```

Comment:
For what it's worth, this string representation is probably more appropriate for the `__str__` function (which I'm guessing would still do what you want).

A `__repr__` function generally tries to give you text that will reproduce the object, so in this case maybe something more like

`return "PackageSet(label=%s, packages=%s, selinux=%s)" % (self.label, self.packages, self.selinux)`

[https://docs.python.org/2/reference/datamodel.html#object.\_\_repr\_\_](https://docs.python.org/2/reference/datamodel.html#object.__repr__)
[https://docs.python.org/2/reference/datamodel.html#object.\_\_str\_\_](https://docs.python.org/2/reference/datamodel.html#object.__str__)
http://stackoverflow.com/questions/1436703/difference-between-str-and-repr-in-python